rule/nested-structs.go

```diff
@@ -0,0 +1,62 @@
+package rule
+
+import (
+	"fmt"
+	"go/ast"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// NestedStructs lints nested structs.
+type NestedStructs struct{}
+
+// Apply applies the rule to given file.
+func (r *NestedStructs) Apply(file *lint.File, arguments lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	if len(arguments) > 0 {
+		panic(r.Name() + " doesn't take any arguments")
+	}
+
+	walker := &lintNestedStructs{
+		fileAST: file.AST,
+		onFailure: func(failure lint.Failure) {
+			failures = append(failures, failure)
+		},
+	}
+
+	ast.Walk(walker, file.AST)
+
+	if walker.count > 0 {
+		walker.onFailure(lint.Failure{
+			Failure:    fmt.Sprintf("no nested structs are allowed, got %d", walker.count),
+			Confidence: 1,
+			Node:       file.AST,
```

Comment:
During the analysis we have enough information to spot the exact struct that do not respect the rule, therefore we can provide a more precise failure, i.e. a failure attached to the struct and not to the whole file.

Imagine a file with hundreds of lines and one nested struct declaration, as user I expect `revive` to point me the single offensive struct and not the whole file.

Reply:
Good catch. I put that there as I was learning how this worked and forgot to change it.